### PR TITLE
K8SPSMDB-321: Add readiness/liveness probe for mongos

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-320](https://jira.percona.com/browse/K8SPSMDB-320)

Should be merged only after this PR https://github.com/percona/mongodb-orchestration-tools/pull/248

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| operator-self-healing | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | failed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| security-context | passed |
| arbiter | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by hors — COMMENTED** [review]
@blez  @delgod  in spec we have for readiness probe: 
```
      failureThreshold: 3
      periodSeconds: 1
```

in code we have:
```
failureThreshold: 8
periodSeconds: 3
```

Is it ok?

**Comment by delgod** [comment]
> @blez @delgod in spec we have for readiness probe:
> 
> ```
>       failureThreshold: 3
>       periodSeconds: 1
> ```
> 
> in code we have:
> 
> ```
> failureThreshold: 8
> periodSeconds: 3
> ```
> 
> Is it ok?

let's follow spec

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | failed |
| users | passed |

**Comment by blez** [comment]
@hors @delgod 
I would say `periodSeconds: 1` is too frequently and suggest to use 
```
failureThreshold: 3
periodSeconds: 5
```

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-479-e3ff8b3c

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| upgrade | passed |
| one-pod | passed |
| init-deploy | passed |
| monitoring | passed |
| self-healing | passed |
| upgrade-consistency | passed |
| limits | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| demand-backup-sharded | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-479-50f90f7f

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| arbiter | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-479-5b548e15

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

File: pkg/psmdb/mongos.go [hunk 1]
```diff
@@ -134,6 +134,12 @@ func mongosContainer(cr *api.PerconaServerMongoDB) (corev1.Container, error) {
 				ContainerPort: cr.Spec.Sharding.Mongos.Port,
 			},
 		},
+		Env: []corev1.EnvVar{
```
Comment (delgod) [inline pkg/psmdb/mongos.go:137]: can we please please add
```
--- e2e-tests/demand-backup-sharded/run
+++ e2e-tests/demand-backup-sharded/run
@@ -26,6 +26,7 @@ wait_for_running $cluster-rs0 3
 desc 'check if service and statefulset created with expected config'
 compare_kubectl statefulset/$cluster-rs0
 compare_kubectl statefulset/$cluster-cfg
+compare_kubectl deploy/$cluster-mongos

 desc 'write data, read from all'
 run_mongos \
```
line into `e2e-tests/demand-backup-sharded/run`
reason: it is needed to see if such changes modify finial structure in k8s in fact

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by hors** [comment]
> @hors @delgod
> I would say `periodSeconds: 1` is too frequently and suggest to use
> 
> ```
> failureThreshold: 3
> periodSeconds: 5
> ```

we need to check it every sec, It is not ok if we detect failed state only in 15 secs

File: pkg/psmdb/mongos.go (hunk 1 above)
Comment (blez) [reply pkg/psmdb/mongos.go:137]: https://github.com/percona/percona-server-mongodb-operator/pull/479/commits/e0596be68f275c17289272098117dd3cb4e6f32d

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| demand-backup-sharded | failed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-479-ac57babd

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| arbiter | passed |
| security-context | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-479-e29f721b

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| one-pod | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| liveness | passed |
| scheduled-backup | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| demand-backup-sharded | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| self-healing | passed |
| limits | passed |
| operator-self-healing | passed |
| scaling | passed |
| security-context | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| one-pod | passed |
| demand-backup-sharded | failed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| arbiter | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| liveness | passed |
| smart-update | passed |
| demand-backup-sharded | failed |
| version-service | passed |
| users | passed |

**Review by delgod — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |

**Review by qJkee — DISMISSED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| monitoring-2-0 | failed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | passed |
| smart-update | failed |
| version-service | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |

**Pull request merged**